### ISSUE TITLE
[pico_protocol] - pico_proto_###.alloc(...) has a dead end at the network layer. 

### ISSUE THREAD

**Issue by jelledevleeschouwer** [tass-belgium/picotcp#405]
### Issue
I've noticed that the procedure of allocating frame as described in the [wiki](https://github.com/tass-belgium/picotcp/wiki/protocols#protocol-output-diagram), higher level modules call following lower level modules' alloc-function, isn't consistent throughout the stack. The fact that it's only applied at the network layer isn't an issue, but the fact that the alloc-chain already stops there, is. If other future lower level protocols apart from ```Ethernet```, like ```IEEE802.15.4``` for example, want to prepend an encapsulation header, there won't be enough overhead available to allow this.

In addition, as I mentioned in issue #376, the 'overhead'-field of the ```pico_device```, doesn't seem to be in actual use. I've grepped for ```"\.overhead"```, and it seems to be set in some modules, but it doesn't seem to be used for allocation anywhere.

### Possible Solution
A possible solution for both of this issues is to update the signature of the pico_protocol's alloc function to: 
```C
struct pico_frame *(*alloc)(struct pico_protocol *self, uint16_t size, struct pico_device *dev);
```
This allows passing the device, based on which the link layer protocol can be chosen and  the appropriate lower level protocol's alloc-function can be called. In the lowest level protocol, probably the link layer, eventually the overhead-member of the device can be used to alloc a frame with enough space for every layer's header.

### Considerations
It has to be possible for a protocol to pass the device to a alloc-function. This is immediately  possible for each of the following modules:
- ```pico_icmp6```
- ```pico_mld```
- ```pico_igmp```

And following unit tests:
- ```modunit_pico_igmp```
- ```modunit_pico_mld```

It's also possible in following modules after some modifications:
- ```pico_socket```
- ```pico_icmp4```

What we could do for example for situations where it isn't directly possible to pass the device, is pass ```NULL```. The callee (alloc-function) must allocate the default frame size, which is the passed size + Ethernet-overhead in most cases. This provides compatibility with how things are happening right now.

Pull-request will follow with modifications.

**Comment by frederikvs** [comment]
Just a couple of questions : 

If in the case of e.g. icmp we pass `NULL` and we allocate for ethernet, does that mean that we can't do icmp over a device with different overhead? (e.g. 802.15.4, but also VLAN ethernet)

Do we have situations where a frame's device is modified along the way? Do we ever clone a frame, then modify the target device and send it on its way? Does NAT just modify some fields and change the device, and send it on its way? Does a broadcast packet get sent out on multiple devices without much modification? 
Because those are situations where this way of working would break...

**Comment by jelledevleeschouwer** [comment]
This is indeed where things seem to get a little tricky. For the notify function I think we can pass the device from the frame that triggered the notification, like [ICMP6 does](https://github.com/jelledevleeschouwer/picotcp/blob/master/modules/pico_icmp6.c#L211). Since, a notification will in most cases be the result of a received frame, and the device will always be set in such a case. I'm not sure if it's the result of a received frame in *all* cases, I think it is, if someone with ICMP knowledge can verify this, we're good.

For the ICMP4-ping, we could implement the [ICMPv4 ping](https://github.com/tass-belgium/picotcp/blob/master/modules/pico_icmp4.c#L337) again like the [ICMPv6 ping](https://github.com/tass-belgium/picotcp/blob/master/modules/pico_icmp4.c#L337) passing the device as well. This involves a lot of rework and rework of example applications, though... 

For TCP, all the ```NULL``` can be changed to ```get_sock_dev``` since TCP sockets will always have a remote address.

**Comment by jelledevleeschouwer** [comment]
@frederikvs For the latter situations you mentioned, I think new link layer protocols should calculate how much headroom there is available before prepending their headers, e.g. subtracting ```f->buffer``` from ```f->net_hdr```. If there's not enough headroom available, the new link layer reallocates the frame once before prepending it's headers, it's not really efficient but I think there's no way around if we want to ensure backwards compatibility. I think the alloc-chain will have provided enough headroom in the majority of the situations, and checking the headroom provides a fail-safe mechanism. 

This check should also be implemented in the Ethernet layer then.

**Comment by jelledevleeschouwer** [comment]
Already provided possibility to grow the headroom of a ```pico_frame``` in commit [c36d44d](https://github.com/tass-belgium/picotcp/pull/406/commits/c36d44de9d8ff2c8e918b413dc100bc26c905578) (#406), just like the ```pico_frame_grow```-function. Sufficient unit test is included in commit [c3b253a](https://github.com/tass-belgium/picotcp/pull/406/commits/c3b253a4b9bdfc7717282e447337502fdd2c0af0). 

This is required, even if we decide not to integrate the change in the alloc-procedure. The change in the alloc-procedure reduces the chance a reallocation/copy is needed, which is more efficient.

**Comment by frederikvs** [comment]
I think this issue was closed by PR #406, correct me if I'm wrong (and reopen)

**Comment by jelledevleeschouwer** [comment]
Check!
